lambdas/tests/unit/services/test_upload_document_reference_service.py

```diff
@@ -662,3 +718,152 @@ def test_process_preliminary_document_reference_exception_during_processing(
+
+
+def test_get_infrastructure_for_document_key_pdm(service):
+    assert service.table_name == MOCK_LG_TABLE_NAME
+    assert service.destination_bucket_name == MOCK_LG_BUCKET
+    service._get_infrastructure_for_document_key(
+        object_parts=["fhir_upload", SnomedCodes.PATIENT_DATA.value.code, "1234"]
+    )
+    assert service.table_name == MOCK_PDM_TABLE_NAME
+    assert service.destination_bucket_name == MOCK_PDM_BUCKET
+
+
+def test_get_infrastructure_for_document_key_non_pdm(service):
+    infra = service._get_infrastructure_for_document_key(object_parts=["1234", "123"])
+    assert infra is None
+
+
+def test_get_infra_invalid_doc_type(monkeypatch, service):
+    # Create a fake doc_type object
+    fake_doc_type = Mock()
+    fake_doc_type.code = "999999"
+    fake_doc_type.display_name = "Fake Doc"
+
+    # Mock SnomedCodes.find_by_code so doc_type is NOT None
+    monkeypatch.setattr(
+        "services.upload_document_reference_service.SnomedCodes.find_by_code",
+        lambda code: fake_doc_type,
+    )
+
+    # Mock routers
+    mock_table_router = Mock()
+    mock_table_router.resolve.side_effect = KeyError("nope")
+
+    mock_bucket_router = Mock()
+
+    # Force KeyError inside the try block
+    mock_table_router.resolve.side_effect = KeyError("not found")
+    service.table_router = mock_table_router
+    service.bucket_router = mock_bucket_router
+    # Call function and assert the exception is raised
+    with pytest.raises(InvalidDocTypeException):
+        service._get_infrastructure_for_document_key(["fhir_upload", "999999"])
+
+
+@pytest.mark.parametrize(
+    "object_key,expected_document_key,expected_table",
+    [
+        (
+            "staging/documents/test-doc-123",
+            "test-doc-123",
+            MOCK_LG_TABLE_NAME,
+        ),
+        ("folder/subfolder/another-doc", "another-doc", MOCK_LG_TABLE_NAME),
+        ("simple-doc", "simple-doc", MOCK_LG_TABLE_NAME),
+        (
+            f"fhir_upload/{SnomedCodes.PATIENT_DATA.value.code}/staging/test-doc-123",
+            "test-doc-123",
+            MOCK_PDM_TABLE_NAME,
+        ),
+        (
+            f"{SnomedCodes.LLOYD_GEORGE.value.code}/staging/test-doc-123",
+            "test-doc-123",
+            MOCK_LG_TABLE_NAME,
+        ),
+        (
+            f"fhir_upload/{SnomedCodes.LLOYD_GEORGE.value.code}/staging/test-doc-123",
+            "test-doc-123",
+            MOCK_LG_TABLE_NAME,
+        ),
+    ],
+)
+def test_document_type_extraction_from_object_key(
+    service,
+    mock_document_reference,
+    object_key,
+    expected_document_key,
+    expected_table,
+):
+    """Test extraction of a document key from various object key formats"""
+    service.document_service.fetch_documents_from_table.return_value = [
+        mock_document_reference
+    ]
+
+    service.handle_upload_document_reference_request(object_key)
+
+    assert service.table_name == expected_table
+
+    # if expected_table != MOCK_LG_TABLE_NAME:
```

Comment:
I want to add these back in at some point but they require a little more though than I had at 7:30 on Friday evening. So I'll fix them at a later stage. I think there needs to be a ticket after this to really fix the problem with agnostic code rather than this branching logic